src/x86/runtime.rs

```diff
@@ -0,0 +1,121 @@
+use ::std::sync::atomic::{AtomicUsize, Ordering};
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[macro_export]
+macro_rules! __unstable_detect_feature {
+    ("sse") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse{}) };
+    ("sse2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse2{}) };
+    ("sse3") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse3{}) };
+    ("ssse3") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::ssse3{}) };
+    ("sse4.1") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse4_1{}) };
+    ("sse4.2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::sse4_2{}) };
+    ("avx") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::avx{}) };
+    ("avx2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::avx2{}) };
+    ("fma") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::fma{}) };
+    ("bmi") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::bmi{}) };
+    ("bmi2") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::bmi2{}) };
+    ("abm") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::abm{}) };
+    ("lzcnt") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::abm{}) };
+    ("tbm") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::tbm{}) };
+    ("popcnt") => { $crate::vendor::__unstable_detect_feature($crate::vendor::__Feature::popcnt{}) };
+}
+
+/// X86 CPU Feature
+#[doc(hidden)]  // Please do not use this, it is an implementation detail subject to change.
+#[allow(non_camel_case_types)]
+#[repr(u8)]
+pub enum __Feature {
+    /// SSE (Streaming SIMD Extensions)
+    sse = 0,
```

Comment:
Are these discriminants necessary?

Reply:
Which ones do you mean? Only the SSE or all of them ?

Reply:
All of them

Reply:
I store whether the feature is detected in an `AtomicUsize`, where each bit represents a feature. This `enum` is used to map feature "names" to bit positions. (I should document this). 

So... I don't understand what you mean. I need to map bit positions to names somewhere. Are you suggesting that I remove the `enum` and use raw numbers for this? Or are you suggesting that I remove the concrete values, that is, `= 3` and use the ones assigned automatically?

Reply:
I've documented the whole module as extensively as I could.

Reply:
Sorry what I meant was that the explicit `= n` for the discriminant isn't necessary here, and I believe it can be removed (the names can and should stay)

Reply:
@alexcrichton If I do that, am I guranteed that the values rustc will assign these discriminants will be in the range [0, 32] ? (I am using `enum as usize` to index the bitset).

Reply:
Indeed you are!

Reply:
Ah I did not know this. I thought that if I did not assign the values then rustc was able to use whatever ordering it wants. Is the sentinel value also defined ? I will just remove the `=`s then.